tools/devservices_healthcheck.py

```diff
@@ -1,47 +1,183 @@
+            f"{zookeeper_origin}:2181",
+            "--list",
+        ),
+        check=True,
+    )
+
+
+def check_postgres():
+    subprocess.run(
+        ("docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"), check=True
+    )
+
+
+# Available health checks
+all_service_healthchecks = {
+    "postgres": HealthCheck(
+        "postgres",
+        "sentry_postgres",
+        True,
+        lambda: subprocess.run(
+            ["docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"], check=True
+        ),
+    ),
+    "kafka": HealthCheck(
+        "kafka",
+        "sentry_kafka",
+        os.getenv("NEED_KAFKA") == "true",
+        check_kafka,
+        deps=["zookeeper"],
+    ),
+    "zookeeper": HealthCheck(
+        "zookeeper",
+        "sentry_zookeeper",
+        os.getenv("NEED_KAFKA") == "true",
+    ),
+}
 
-        if returncode != 0:
-            time.sleep(timeout)
+
+def run_with_retries(cmd: Callable[[], object], retries: int, timeout: int) -> None:
+    for retry in range(1, retries + 1):
+        try:
+            cmd()
+        except Exception as e:
+            if retry == retries:
+                print(f"Command failed, no more retries: {e}")
+                raise HealthcheckError(f"Command failed: {e}")
+            else:
+                print(f"Command failed, retrying in {timeout}s (attempt {retry+1} of {retries})...")
+                time.sleep(timeout)
         else:
             return
 
-    raise SystemExit(1)
 
+def get_services_to_check(id: str) -> list[str]:
+    if all_service_healthchecks.get(id) is None:
+        raise HealthcheckError(f"Service '{id}' does not have a health check")
 
-def main() -> None:
-    # Available health checks
-    postgres_healthcheck = ["docker", "exec", "sentry_postgres", "pg_isready", "-U", "postgres"]
-    kafka_healthcheck = [
-        "docker",
-        "exec",
-        "sentry_kafka",
-        "kafka-topics",
-        "--zookeeper",
-        # TODO: sentry_zookeeper:2181 doesn't work in CI, but 127.0.0.1 doesn't work locally
-        "127.0.0.1:2181",
-        "--list",
-    ]
-
-    healthchecks = [postgres_healthcheck]
-    if os.getenv("NEED_KAFKA") == "true":
-        healthchecks.append(kafka_healthcheck)
-
-    for check in healthchecks:
-        run_cmd(check)
+    checks = []
+    hc = all_service_healthchecks[id]
+    for dep in hc.deps:
+        dep_checks = get_services_to_check(dep)
+        for d in dep_checks:
+            checks.append(d)
+    checks.append(id)
+    return checks
+
+
+def check_health(ids: list[str]) -> None:
+    checks = []
+    for id in ids:
+        s = get_services_to_check(id)
+        checks += s
+
+    # dict.fromkeys is used to remove duplicates while maintaining order
+    for name in dict.fromkeys(checks):
+        print(f"Checking service {name}")
+        hc = all_service_healthchecks[name]
+        print(f"Checking '{hc.container_name}' is running...")
+        ls = " ".join(list(set(checks)))
```

Comment:
this removes the ordering invariant you established above -- maybe assign that and reuse it here?